mist-core/src/main/java/edu/snu/mist/core/master/SingleNodeRecoveryScheduler.java

```
@@ -0,0 +1,148 @@
+   */
+  private boolean isRecovering;
+
+  @Inject
+  private SingleNodeRecoveryScheduler(
+      final TaskStatsMap taskStatsMap,
+      final ProxyToTaskMap proxyToTaskMap) {
+    this.allFailedGroups = new ConcurrentHashMap<>();
+    this.recoveryGroups = null;
+    this.taskStatsMap = taskStatsMap;
+    this.proxyToTaskMap = proxyToTaskMap;
+    this.lock = new ReentrantLock();
+    this.notInRecoveryProcess = this.lock.newCondition();
+    this.isRecovering = false;
+  }
+
+  @Override
+  public synchronized void addFailedGroups(final Map<String, GroupStats> failedGroups) {
+    this.allFailedGroups.putAll(failedGroups);
+  }
+
+  @Override
+  public synchronized void startRecovery() {
+    try {
+      lock.lock();
+      while (isRecovering) {
+        notInRecoveryProcess.await();
+      }
+      isRecovering = true;
+      this.recoveryGroups = allFailedGroups;
+      this.allFailedGroups = new ConcurrentHashMap<>();
+      // Select the newly allocated task with the least load for recovery
+      double minLoad = Double.MAX_VALUE;
+      proxyToRecoveryTask = null;
+      for (final Map.Entry<String, MasterToTaskMessage> entry : proxyToTaskMap.entrySet()) {
+        final String taskHostname = entry.getKey();
+        final double taskLoad = taskStatsMap.get(taskHostname).getTaskLoad();
+        if (taskLoad < minLoad) {
+          minLoad = taskLoad;
+          proxyToRecoveryTask = entry.getValue();
+        }
+      }
+      proxyToRecoveryTask.startRecovery();
+    } catch (final InterruptedException e) {
+      LOG.log(Level.SEVERE, "Recovery has been interrupted. " + e.toString());
+    } catch (final AvroRemoteException e) {
+      LOG.log(Level.SEVERE, "Start recovery through avro server has failed! " + e.toString());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public List<String> getRecoveringGroups(final String taskHostname) {
+    // No more groups to be recovered! Recovery is done!
+    if (recoveryGroups.isEmpty()) {
```

`recoveryGroups` can be null if it is called before `startRecovery()`. At least, we should throw an exception .